File: app/src/main/java/com/kapasiya/sharefair/SignUpActivity.java
package com.kapasiya.sharefair;

import static android.content.ContentValues.TAG;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.util.Patterns;
import android.widget.CheckBox;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.credentials.CredentialManager;
import androidx.credentials.CredentialManagerCallback;
import androidx.credentials.GetCredentialRequest;
import androidx.credentials.GetCredentialResponse;
import androidx.credentials.exceptions.GetCredentialException;

import com.bumptech.glide.Glide;
import com.google.android.libraries.identity.googleid.GetGoogleIdOption;
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential;
import com.google.android.material.button.MaterialButton;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.AuthCredential;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseAuthUserCollisionException;
import com.google.firebase.auth.FirebaseAuthWeakPasswordException;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.auth.GoogleAuthProvider;
import com.google.firebase.auth.UserProfileChangeRequest;

public class SignUpActivity extends AppCompatActivity {

    private TextInputEditText fullName;
    private TextInputEditText email;
    private TextInputEditText password;
    private TextInputEditText cPassword;
    private MaterialButton createAccount;
    private MaterialButton googleLoginButton;
    private CheckBox termsCheckBox;

    private CredentialManager credentialManager;
    private FirebaseAuth auth;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_signup);

        // Initialize Firebase Auth
        auth = FirebaseAuth.getInstance();

        // Initialize Credential Manager (modern replacement for Google Sign-In)
        credentialManager = CredentialManager.create(this);

        initializeViews();
        setupClickListeners();
    }

    private void initializeViews() {
        googleLoginButton = findViewById(R.id.googleSignUpButton);
        createAccount = findViewById(R.id.createAccountButton);
        fullName = findViewById(R.id.fullNameSignUp);
        email = findViewById(R.id.emailSignUp);
        password = findViewById(R.id.passwordSignup);
        cPassword = findViewById(R.id.cPasswordSignup);
        termsCheckBox = findViewById(R.id.termsCheckBox);
    }

    private void setupClickListeners() {
        googleLoginButton.setOnClickListener(view -> signInWithGoogle());

        createAccount.setOnClickListener(view -> createAccountWithEmail());

        findViewById(R.id.signInText).setOnClickListener(view -> {
            Intent intent = new Intent(SignUpActivity.this, LoginActivity.class);
            startActivity(intent);
            finish();
        });

        findViewById(R.id.backButton).setOnClickListener(view -> {
            Intent intent = new Intent(SignUpActivity.this, LoginActivity.class);
            startActivity(intent);
            finish();
        });

        findViewById(R.id.termsText).setOnClickListener(view -> {
            // Open terms and conditions
            Toast.makeText(this, "Terms & Conditions clicked", Toast.LENGTH_SHORT).show();
            // You can open a WebView or another activity with terms
        });
    }

    private void createAccountWithEmail() {
        String fullNameText = fullName.getText().toString().trim();
        String emailText = email.getText().toString().trim();
        String passwordText = password.getText().toString().trim();
        String cPasswordText = cPassword.getText().toString().trim();

        // Validate inputs
        if (!validateInputs(fullNameText, emailText, passwordText, cPasswordText)) {
            return;
        }

        // Check terms and conditions
        if (!termsCheckBox.isChecked()) {
            Toast.makeText(this, "Please accept Terms & Conditions", Toast.LENGTH_SHORT).show();
            return;
        }

        // Disable button to prevent multiple clicks
        createAccount.setEnabled(false);
        createAccount.setText("Creating Account...");

        // Create user with email and password
        auth.createUserWithEmailAndPassword(emailText, passwordText)
                .addOnCompleteListener(this, task -> {
                    createAccount.setEnabled(true);
                    createAccount.setText("Create Account");

                    if (task.isSuccessful()) {
                        Log.d(TAG, "createUserWithEmail:success");
                        FirebaseUser user = auth.getCurrentUser();

                        if (user != null) {
                            // Update user profile with full name
                            UserProfileChangeRequest profileUpdates = new UserProfileChangeRequest.Builder()
                                    .setDisplayName(fullNameText)
                                    .build();

                            user.updateProfile(profileUpdates)
                                    .addOnCompleteListener(profileTask -> {
                                        if (profileTask.isSuccessful()) {
                                            Log.d(TAG, "User profile updated.");
                                        }
                                    });

                            // Send email verification
                            sendEmailVerification(user, emailText);
                        }
                    } else {
                        Log.w(TAG, "createUserWithEmail:failure", task.getException());
                        handleSignUpError(task.getException());
                    }
                });
    }

    private boolean validateInputs(String fullName, String email, String password, String cPassword) {
        if (fullName.isEmpty()) {
            this.fullName.setError("Full name is required");
            this.fullName.requestFocus();
            return false;
        }

        if (email.isEmpty()) {
            this.email.setError("Email is required");
            this.email.requestFocus();
            return false;
        }

        if (!Patterns.EMAIL_ADDRESS.matcher(email).matches()) {
            this.email.setError("Please enter a valid email");
            this.email.requestFocus();
            return false;
        }

        if (password.isEmpty()) {
            this.password.setError("Password is required");
            this.password.requestFocus();
            return false;
        }

        if (password.length() < 6) {
            this.password.setError("Password must be at least 6 characters");
            this.password.requestFocus();
            return false;
        }

        if (!password.equals(cPassword)) {
            this.cPassword.setError("Passwords do not match");
            this.cPassword.requestFocus();
            return false;
        }

        return true;
    }

    private void sendEmailVerification(FirebaseUser user, String emailText) {
        user.sendEmailVerification()
                .addOnCompleteListener(task -> {
                    if (task.isSuccessful()) {
                        Toast.makeText(SignUpActivity.this,
                                "Account created! Please check your email for verification.",
                                Toast.LENGTH_LONG).show();

                        // Navigate to OTP verification activity
                        Intent intent = new Intent(SignUpActivity.this, EmailVerificationActivity.class);
                        intent.putExtra("email", emailText);
                        startActivity(intent);
                        finish();
                    } else {
                        Toast.makeText(SignUpActivity.this,
                                "Failed to send verification email: " +
                                        (task.getException() != null ? task.getException().getMessage() : "Unknown error"),
                                Toast.LENGTH_SHORT).show();
                        Log.e(TAG, "Failed to send verification email", task.getException());
                    }
                });
    }

    private void handleSignUpError(Exception exception) {
        if (exception instanceof FirebaseAuthWeakPasswordException) {
            password.setError("Password is too weak");
            password.requestFocus();
        } else if (exception instanceof FirebaseAuthUserCollisionException) {
            email.setError("An account with this email already exists");
            email.requestFocus();
        } else {
            Toast.makeText(this, "Sign up failed: " +
                            (exception != null ? exception.getMessage() : "Unknown error"),
                    Toast.LENGTH_SHORT).show();
        }
    }

    private void signInWithGoogle() {
        // Create GetGoogleIdOption (replacement for GoogleSignInOptions)
        GetGoogleIdOption googleIdOption = new GetGoogleIdOption.Builder()
                .setFilterByAuthorizedAccounts(false)
                .setServerClientId(getString(R.string.client_id))
                .setAutoSelectEnabled(true)
                .build();

        // Create GetCredentialRequest
        GetCredentialRequest request = new GetCredentialRequest.Builder()
                .addCredentialOption(googleIdOption)
                .build();

        // Get credential using CredentialManager
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            credentialManager.getCredentialAsync(
                    this,
                    request,
                    null, // CancellationSignal - can be null
                    this.getMainExecutor(), // Executor
                    new CredentialManagerCallback<GetCredentialResponse, GetCredentialException>() {
                        @Override
                        public void onResult(GetCredentialResponse result) {
                            handleGoogleSignInResult(result);
                        }

                        @Override
                        public void onError(GetCredentialException e) {
                            Log.w(TAG, "Google Sign-In failed", e);
                            Toast.makeText(SignUpActivity.this, "Google Sign-In failed: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                        }
                    }
            );
        }
    }

    @SuppressLint("CheckResult")
    private void handleGoogleSignInResult(GetCredentialResponse result) {
        GoogleIdTokenCredential googleIdTokenCredential = GoogleIdTokenCredential
                .createFrom(result.getCredential().getData());

        String idToken = googleIdTokenCredential.getIdToken();

        // Create Firebase credential
        AuthCredential firebaseCredential = GoogleAuthProvider.getCredential(idToken, null);

        // Sign in to Firebase
        auth.signInWithCredential(firebaseCredential)
                .addOnCompleteListener(this, task -> {
                    if (task.isSuccessful()) {
                        Log.d(TAG, "signInWithCredential:success");

                        // Load user profile image
                        if (auth.getCurrentUser() != null && auth.getCurrentUser().getPhotoUrl() != null) {
                            Glide.with(SignUpActivity.this).load(auth.getCurrentUser().getPhotoUrl());
                        }

                        Log.w(TAG, "Name: " + (auth.getCurrentUser() != null ? auth.getCurrentUser().getDisplayName() : "null"));
                        Log.w(TAG, "Mail: " + (auth.getCurrentUser() != null ? auth.getCurrentUser().getEmail() : "null"));

                        Toast.makeText(SignUpActivity.this, "Signed in successfully!", Toast.LENGTH_SHORT).show();
                        navigateToMainActivity();
                    } else {
                        Log.w(TAG, "signInWithCredential:failure", task.getException());
                        Toast.makeText(SignUpActivity.this, "Authentication failed: " +
                                        (task.getException() != null ? task.getException().getMessage() : "Unknown error"),
                                Toast.LENGTH_SHORT).show();
                    }
                });
    }

    private void navigateToMainActivity() {
        Intent intent = new Intent(SignUpActivity.this, DashboardActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        startActivity(intent);
        finish();
    }
}

File: app/src/main/java/com/kapasiya/sharefair/SignUpActivity.kt
package com.kapasiya.sharefair

import android.content.ContentValues.TAG
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.credentials.CredentialManager
import androidx.credentials.GetCredentialRequest
import androidx.credentials.GetCredentialResponse
import androidx.credentials.exceptions.GetCredentialException
import androidx.credentials.exceptions.NoCredentialException
import androidx.core.content.ContextCompat
import androidx.lifecycle.lifecycleScope
import com.google.android.libraries.identity.googleid.GetGoogleIdOption
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.FirebaseAuthUserCollisionException
import com.google.firebase.auth.FirebaseAuthWeakPasswordException
import com.google.firebase.auth.FirebaseUser
import com.google.firebase.auth.GoogleAuthProvider
import com.google.firebase.auth.UserProfileChangeRequest
import com.kapasiya.sharefair.ui.theme.ShareFairTheme
import com.kapasiya.sharefair.ui.screens.SignUpScreen
import com.kapasiya.sharefair.model.Group
import com.kapasiya.sharefair.model.User
import com.kapasiya.sharefair.data.RepositoryModule
import kotlinx.coroutines.launch
import java.security.MessageDigest
import java.util.UUID

class SignUpActivity : ComponentActivity() {

    private lateinit var credentialManager: CredentialManager
    private lateinit var auth: FirebaseAuth

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        auth = FirebaseAuth.getInstance()
        credentialManager = CredentialManager.create(this)

        setContent {
            ShareFairTheme {
                SignUpScreen(
                    onSignUp = { name, email, password -> 
                        createAccountWithEmail(name, email, password)
                    },
                    onGoogleSignIn = { signInWithGoogle() },
                    onSignIn = {
                        startActivity(Intent(this, LoginActivity::class.java))
                        finish()
                    }
                )
            }
        }
    }

    private fun createAccountWithEmail(fullNameText: String, emailText: String, passwordText: String) {
        auth.createUserWithEmailAndPassword(emailText, passwordText)
            .addOnCompleteListener(this) { task ->
                if (task.isSuccessful) {
                    val firebaseUser = auth.currentUser
                    firebaseUser?.let {
                        val profileUpdates = UserProfileChangeRequest.Builder()
                            .setDisplayName(fullNameText)
                            .build()
                        it.updateProfile(profileUpdates)
                        
                        // Create user profile in DB immediately
                        syncUserWithDatabase(it, fullNameText)
                        sendEmailVerification(it, emailText)
                    }
                } else {
                    handleSignUpError(task.exception)
                }
            }
    }

    private fun sendEmailVerification(user: FirebaseUser, emailText: String) {
        user.sendEmailVerification().addOnCompleteListener { task ->
            if (task.isSuccessful) {
                val intent = Intent(this, EmailVerificationActivity::class.java)
                intent.putExtra("email", emailText)
                startActivity(intent)
                finish()
            }
        }
    }

    private fun handleSignUpError(exception: Exception?) {
        val message = when (exception) {
            is FirebaseAuthWeakPasswordException -> "Password is too weak"
            is FirebaseAuthUserCollisionException -> "Account already exists"
            else -> "Sign up failed: ${exception?.message}"
        }
        Toast.makeText(this, message, Toast.LENGTH_SHORT).show()
    }

    private fun signInWithGoogle() {
        val serverClientId = getString(R.string.default_web_client_id)
        val rawNonce = UUID.randomUUID().toString()
        val bytes = rawNonce.toByteArray()
        val md = MessageDigest.getInstance("SHA-256")
        val digest = md.digest(bytes)
        val hashedNonce = digest.fold("") { str, it -> str + "%02x".format(it) }

        val googleIdOption = GetGoogleIdOption.Builder()
            .setFilterByAuthorizedAccounts(false)
            .setServerClientId(serverClientId)
            .setNonce(hashedNonce)
            .setAutoSelectEnabled(false)
            .build()

        val request = GetCredentialRequest.Builder().addCredentialOption(googleIdOption).build()

        val executor = ContextCompat.getMainExecutor(this)
        credentialManager.getCredentialAsync(this, request, null, executor,
            object : androidx.credentials.CredentialManagerCallback<GetCredentialResponse, GetCredentialException> {
                override fun onResult(result: GetCredentialResponse) {
                    try {
                        val credential = GoogleIdTokenCredential.createFrom(result.credential.data)
                        auth.signInWithCredential(GoogleAuthProvider.getCredential(credential.idToken, null))
                            .addOnCompleteListener { task ->
                                if (task.isSuccessful) {
                                    val firebaseUser = auth.currentUser
                                    if (firebaseUser != null) {
                                        syncUserWithDatabase(firebaseUser)
                                    } else {
                                        navigateToMainActivity()
                                    }
                                } else {
                                    Log.e(TAG, "Firebase auth failed", task.exception)
                                    Toast.makeText(this@SignUpActivity, "Google Sign-in failed", Toast.LENGTH_LONG).show()
                                }
                            }
                    } catch (e: Exception) {
                        Log.e(TAG, "Google ID Token parsing failed", e)
                    }
                }
                override fun onError(e: GetCredentialException) {
                    Log.e(TAG, "Auth failed: ${e.javaClass.simpleName}", e)
                }
            }
        )
    }

    private fun syncUserWithDatabase(firebaseUser: FirebaseUser, manualName: String? = null) {
        val userRepo = RepositoryModule.userRepository
        val groupRepo = RepositoryModule.groupRepository
        
        lifecycleScope.launch {
            try {
                val existingProfile = userRepo.getUserProfile(firebaseUser.uid)
                if (existingProfile == null) {
                    val newProfile = User(
                        id = firebaseUser.uid,
                        name = manualName ?: firebaseUser.displayName ?: "User",
                        email = firebaseUser.email ?: "",
                        profileImageUrl = firebaseUser.photoUrl?.toString() ?: "",
                        totalBalance = 0.0,
                        friends = emptyList(),
                        isPremium = false,
                        groupLimit = 5
                    )
                    userRepo.saveUserProfile(newProfile)
                    
                    // Feature: Auto-create a "Personal" group for individual spending
                    val personalGroup = Group(
                        name = "My Expenses",
                        members = listOf(firebaseUser.uid),
                        type = "PERSONAL",
                        recentActivity = listOf("Welcome to ShareFair! Track your personal spends here.")
                    )
                    groupRepo.createGroup(personalGroup)
                }
                navigateToMainActivity()
            } catch (e: Exception) {
                Log.e(TAG, "Failed to sync user", e)
                navigateToMainActivity()
            }
        }
    }

    private fun navigateToMainActivity() {
        startActivity(Intent(this, DashboardActivity::class.java))
        finish()
    }
}
